### ISSUE TITLE
ambiguity in constructs.lua

### ISSUE THREAD

**Issue by rns** [ronsavage/MarpaX-Languages-Lua-Parser#3]
```
The Lua source is ambiguous: Length of symbol "<stat item>" at line 153, column 1 is ambiguous
  Choices start with: f,a = loadstring
  Choice 1, length=16, ends at line 153, column 16
  Choice 1: f,a = loadstring
  Choice 2, length=22, ends at line 153, column 22
  Choice 2: f,a = loadstring(f)();
```

rules involved:

```
<stat item> ::= <stat>
...
<stat> ::= <functioncall>
...
<exp> ::=
        <var>
     | '(' <exp> ')' assoc => group
    || <exp> <args> assoc => right
    || <exp> ':' <Name> <args> assoc => right
...
<prefixexp> ::= <var>
<prefixexp> ::= <functioncall>
<prefixexp> ::= '(' <exp> ')'

<functioncall> ::= <prefixexp> <args>
<functioncall> ::= <prefixexp> ':' <Name> <args>
```

**Comment by jeffreykegler** [comment]
Hmmm

**Comment by rns** [comment]
I'm remembering using ranks in a similar situation with Lua::AST:

https://github.com/rns/MarpaX-Languages-Lua-AST/blob/master/lib/MarpaX/Languages/Lua/AST.pm#L54

https://github.com/rns/MarpaX-Languages-Lua-AST/blob/master/lib/MarpaX/Languages/Lua/AST.pm#L72

**Comment by jeffreykegler** [comment]
So maybe that fix can be ported across?

**Comment by rns** [comment]
I tried, but it doesn't work, perhaps because I use explicit left recursion where Lua::Parser uses sequences.

**Comment by jeffreykegler** [comment]
Perhaps convert to explicit left recursion, then.

**Comment by rns** [comment]
@jeffreykegler -- ok, done in the patch below; what do you think -- can it be done in bnf?

@ronsavage -- would you like to try it to see how it works with your parse tree processing code? There can be deep recursion warning in traversal, can be silenced with

``` perl
my $DOWARN = 1;
BEGIN { $SIG{'__WARN__'} = sub { warn $_[0] if $DOWARN } }
...
$DOWARN = 0;
        # parse tree traversal code ...
$DOWARN = 1;
```

the patch

```
--- a/lib/MarpaX/Languages/Lua/Parser.pm
+++ b/lib/MarpaX/Languages/Lua/Parser.pm
@@ -749,21 +749,23 @@ lexeme default = latm => 1 action => [name,values]
 # Names which begin with "Lua" are taken directly from
 # the Lua reference manual grammar.

-<chunk> ::= <stat list> <optional laststat>
-
-<stat list> ::= <stat item>*
-<stat item> ::= <stat> ';'
-<stat item> ::= <stat>
-
-<optional laststat> ::= <laststat> ';'
-<optional laststat> ::= <laststat>
-<optional laststat> ::=
+<chunk> ::=
+<chunk> ::= <stat list>
+<chunk> ::= <stat list> laststat
+<chunk> ::= <stat list> laststat ';'
+<chunk> ::= laststat ';'
+<chunk> ::= laststat
+
+<stat list> ::= <stat>
+<stat list> ::= <stat> ';'
+<stat list> ::= <stat list> <stat> rank => -1
+<stat list> ::= <stat list> <stat> ';'

 <block> ::= <chunk>

 <stat> ::= <varlist> '=' <explist>

-<stat> ::= <functioncall>
+<stat> ::= <functioncall> rank => -1

 <stat> ::= <keyword do> <block> <keyword end>
```

**Comment by jeffreykegler** [comment]
@rns -- "Can if be done in BNF?"  We'll see.  It looks like you already did
it in MX::L::Lua::AST.

The sequence notation is simply a shorthand for a left-recursion, and so
when things get complicated, as in this case where the recursion is
ambiguous, writing it direct in BNF is the thing to do.

**Comment by jeffreykegler** [comment]
Here's another (untested) idea.  Create a separate pair of lexemes for the parens which delimit argument lists.  For example, `<largdelim>` and `<rargdelim>`.  Given them a lexeme priority higher than regular parens.

Then, whenever an arglist is expected, the SLIF will not accept a paren that wraps an expression.

Other solutions may be more general, powerful and/or elegant, but I suspect this is a very close match to the parsing logic that Roberto actually wrote.

**Comment by ronsavage** [comment]
I've released V 1.04 incorporating the above.

**Comment by jeffreykegler** [comment]
Does the test suite cover the issue?  Did it go away?

**Comment by rns** [comment]
> Does the test suite cover the issue?

Yes, lua.sources/constructs.lua  -- https://github.com/ronsavage/MarpaX-Languages-Lua-Parser/blob/master/lua.sources/constructs.lua#L153

> Did it go away?

Looks like it, here are the expected tokens -- https://github.com/ronsavage/MarpaX-Languages-Lua-Parser/blob/master/lua.output/constructs.txt#L3677 and the test passes when I just installed it under winxp/perl 5.20.

**Comment by jeffreykegler** [comment]
Cool.  I hope they like it!

**Comment by rns** [comment]
Fixed in just released 1.04
